doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -5,269 +5,291 @@ Reasoning with inductive types
-      between :token:`term` and the value that it takes in each of the
-      possible cases. The name of the equation is specified
-      by :token:`naming_intropattern` (see :tacn:`intros`),
-      in particular ``?`` can be used to let Coq generate a fresh name.
+.. tacn:: destruct {+, @induction_clause } {? @induction_principle }
 
-   .. tacv:: destruct @term with @bindings
+   .. insertprodn induction_clause induction_principle
 
-      This behaves like :n:`destruct @term` providing explicit instances for
-      the dependent premises of the type of :token:`term`.
+   .. prodn::
+      induction_clause ::= @destruction_arg {? as @induction_ipat } {? eqn : @naming_intropattern } {? @occurrences }
+      destruction_arg ::= @natural
+      | @one_term_with_bindings
+      induction_ipat ::= @ident
+      | @or_and_intropattern
+      | @equality_intropattern
+      induction_principle ::= using @one_term {? with @bindings } {? @occurrences }
+
+   Performs case analysis by generating a subgoal for each constructor of the
+   inductive or co-inductive type in the :n:`@one_term` in :n:`@destruction_arg`.
+   In each new subgoal, the tactic replaces :n:`@one_term` with the associated
+   constructor (applied to its arguments).   Unlike :tacn:`induction`,
+   :tacn:`destruct` generates no induction hypothesis.
+
+   The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
+   the tactic checks that all subterms matching the pattern in the conclusion
+   and the hypotheses are compatible and performs case analysis using this subterm.
+
+   .. todo PR Need more explanation about the tactic does when destruction_arg is
+      a one_term that's not just an identifier'
+
+   :n:`{+, @induction_clause }`
+     Giving multiple :n:`@induction_clause`\s is equivalent to applying :n:`destruct`
+     serially on each :n:`@induction_clause`.
+
+   :n:`@natural`
+     Equivalent to :tacn:`intros` :n:`until @natural` followed by `destruct` applied to
+     the last introduced hypothesis.
 
-   .. tacv:: edestruct @term
-      :name: edestruct
+     .. note::
+        To destruct a number, use the syntax :n:`destruct (@natural)` (not
+        very interesting anyway).
 
-      This tactic behaves like :n:`destruct @term` except that it does not
+   :n:`@ident` (in :n:`@destruction_arg`)
+     If :token:`ident` is a quantified variable in the conclusion
+     of the goal, then :n:`destruct @ident` behaves like :tacn:`intros`
+     :n:`until @ident; destruct @ident`.
+     If :token:`ident` is a hypothesis of the context and :token:`ident`
+     is no longer dependent in the goal after application
+     of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as
+     in :n:`destruct (@ident)`).
+
+   :n:`as @induction_ipat`
+      Provides names for the variables introduced
+      into the context.  The :token:`or_and_intropattern` must have the
+      form :n:`[ {*| {* @intropattern } } ]` with one :n:`{* @intropattern }`
```

Comment:
With this formulation, it's not so clear that only this form is accepted (and not an `equality_intropattern` or an `ident`).

Reply:
It appears that `induction_clause` (`as_or_and_ipat` in the mlgs) never accepts `equality_intropattern` (Looking at (e)case, (e)destruct, (e)induction and * inversion*).

IIUC, the change in cf4105502388e437c1cf361b5c3ddd8a482eef04 (allowing `equality_intropattern` in `as_or_and_ipat`) was not needed; the added production is never semantically permitted.  @herbelin?

Also, an `identref` is not accepted for these tactics (in `or_and_intropattern_loc`).  Since the `identref` is only reachable from these tactics, that production can be dropped from the grammar, in which case `or_and_intropattern_loc` can also be dropped, substituting `or_and_intropattern` in `as_or_and_ipat`.

The deleted productions (before dropping `or_and_intropattern_loc`:

or_and_intropattern_loc: [
| or_and_intropattern
~| identref~
]

as_or_and_ipat: [
| "as" or_and_intropattern_loc
~| "as" equality_intropattern~
|
]

I can put in a PR to do this if that's agreeable.

Reply:
> IIUC, the change in [cf41055](https://github.com/coq/coq/commit/cf4105502388e437c1cf361b5c3ddd8a482eef04) (allowing `equality_intropattern` in `as_or_and_ipat`) was not needed; the added production is never semantically permitted.

I get it now! Thanks for digging this commit out. This was as an answer to Ralf in #13205, who was surprised that `destruct` didn't accept the injection intropattern. Therefore, this really is an implementation detail to improve user messages but should not appear in the doc.

Reply:
I simplified the grammar (dropping `induction_ipat`) and reworded the description:

   :n:`as @or_and_intropattern`
      Provides names for the variables introduced in each new subgoal.  The
      :token:`or_and_intropattern` must have one :n:`{* @intropattern }`
      for each constructor, given in the order in which the constructors are
      defined.  If there are not enough names, Coq picks fresh names.